File: Python/Most Frequent Even Element.py
from collections import Counter
from typing import List


class Solution:
    """
    Time:   O(n)
    Memory: O(n)
    """

    def mostFrequentEven(self, nums: List[int]) -> int:
        even_nums = (n for n in nums if not n & 1)
        count = Counter(even_nums)
        return max(count, key=lambda x: (count[x], -x)) if count else -1

File: Java/Most Frequent Even Element.java
import java.util.HashMap;

class Solution {

    /**
     * Time:   O(n)
     * Memory: O(n)
     */
    public int mostFrequentEven(int[] nums) {
        HashMap<Integer, Integer> numFrequency = new HashMap<>();
        int mostFreq = -1, maxFreq = 0;

        for (var n : nums) {
            if (n % 2 != 0)
                continue;

            int freq = numFrequency.getOrDefault(n, 0) + 1;
            numFrequency.put(n, freq);

            if (freq > maxFreq || freq == maxFreq && mostFreq > n) {
                mostFreq = n;
                maxFreq = freq;
            }
        }

        return mostFreq;
    }
}
